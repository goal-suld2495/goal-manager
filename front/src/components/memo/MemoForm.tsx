import { FormEvent, useEffect } from 'react';
import styled from 'styled-components';
import { message } from 'antd';
import { useNavigate } from 'react-router';
import useMemoForm from '../../hooks/useMemoForm';

const MemoFormBlock = styled.form`
  height: 100vh;
  padding: 30px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  .title {
    display: flex;
    flex-direction: column;
    margin-bottom: 30px;
  }
  .title label {
    margin-bottom: 10px;
    font-size: var(--label-font-size);
    color: var(--label-color);
  }
  .title input {
    border: none;
    border-bottom: 1px solid #bbb;
    outline: none;
    transition: border 0.3s ease;
    font-size: var(--label-font-size);
  }
  .title input:focus {
    border-bottom: 1px solid #707070;
  }
  .content {
    flex: 1;
  }
  .content textarea {
    width: 100%;
    height: 100%;
    padding: 30px;
    box-sizing: border-box;
    border: 1px solid #e5e5e5;
    background: #e5e5e5;
    font-size: var(--label-font-size);
    outline: none;
    transition: border 0.3s ease;
    resize: none;
  }
  .content textarea:focus {
    border: 1px solid #a9a9a9;
  }
  .buttons {
    margin-top: 20px;
  }
`;

interface Memo {
  title: string;
  content: string;
}

export type MemoFormProps = {
  initForm?: Memo;
};

const MemoForm = ({ initForm }: MemoFormProps) => {
  const { form, memo, memoError, setForm, saveMemo } = useMemoForm(initForm);
  const navigate = useNavigate();

  const onChange = ({ target }: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const { name, value } = target;

    setForm({
      ...form,
      [name]: value
    });
  };

  const onSubmit = (e: FormEvent) => {
    e.preventDefault();

    if (!form.title || !form.content) {
      message.warning('제목 또는 내용을 입력하세요.');
      return;
    }

    saveMemo();
  };

  useEffect(() => {
    if (memoError) {
      message.error('저장에 실패하였습니다.');
    }
  }, [memoError]);

  useEffect(() => {
    if (memo && memo.id) {
      navigate(`/memos/${memo.id}`);
    }
  }, [memo]);

  return (
    <MemoFormBlock onSubmit={onSubmit}>
      <div className="title">
        <label htmlFor="title">제목</label>
        <input
          id="title"
          name="title"
          type="text"
          value={form.title}
          onChange={onChange}
        />
      </div>
      <div className="content">
        <textarea
          name="content"
          placeholder="내용을 입력해 주세요."
          value={form.content}
          onChange={onChange}
        />
      </div>
      <div className="buttons">
        <button type="submit">생성</button>
      </div>
    </MemoFormBlock>
  );
};

export default MemoForm;
