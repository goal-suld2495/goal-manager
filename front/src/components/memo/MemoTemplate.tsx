import styled from 'styled-components';

const MemoTemplateBlock = styled.div``;

const MemoTemplate = () => {
  return <MemoTemplateBlock />;
};

export default MemoTemplate;
