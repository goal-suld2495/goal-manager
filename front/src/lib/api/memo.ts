import { AxiosResponse } from 'axios';
import apiClient from './apiClient';
import { Form, Memo } from '../../types';

export const saveMemo = (form: Form): Promise<AxiosResponse<Memo>> => {
  return apiClient.post('api/memos', form);
};

export const fetchMemo = (id: string): Promise<AxiosResponse<Memo>> => {
  return Promise.resolve({
      data: { title: '제목', content: '내용', id },
      status: 200,
      statusText: '',
      headers: '',
      config: {},
    });
  // return apiClient.post(`api/memos/${id}`);
};

export const fetchMemoList = (): Promise<AxiosResponse<Memo[]>> => {
  return apiClient.get('api/memos');
};

export default {};
