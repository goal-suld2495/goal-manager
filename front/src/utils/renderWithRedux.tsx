import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { createStore } from 'redux';
import rootReducer, { RootState } from '../modules/store';

function renderWithRedux(
  ui: React.ReactNode,
  initialState: RootState = { memo: { titil: '' } }
) {
  const store = createStore(rootReducer, initialState);
  const utils = render(
    <Provider store={store}>
      <MemoryRouter>{ui}</MemoryRouter>
    </Provider>
  );

  return {
    ...utils,
    store,
  };
}

export default renderWithRedux;
