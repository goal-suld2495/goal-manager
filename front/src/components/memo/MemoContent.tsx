import { useParams } from 'react-router';
import { useRecoilValue } from 'recoil';
import styled from 'styled-components';
import { CurrentMemoState } from '../../modules/memo';
import { Memo } from '../../types';

const MemoContentBlock = styled.div``;

const MemoContent = () => {
  const { id } = useParams<{ id: string }>();

  if (!id) {
    return null;
  }

  let memo: Memo | null;

  try {
    memo = useRecoilValue(CurrentMemoState(id));
  } catch {
    return <div>메모를 불러오는데 실패하였습니다.</div>;
  }

  return (
    <MemoContentBlock>
      <>
        <h1>{memo.title}</h1>
        <div>{memo.content}</div>
      </>
    </MemoContentBlock>
  );
};

export default MemoContent;
