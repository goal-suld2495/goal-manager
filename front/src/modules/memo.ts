import { takeLatest } from 'redux-saga/effects';
import {
  ActionType,
  createAction,
  createAsyncAction,
  createReducer,
} from 'typesafe-actions';
import { AxiosResponse } from 'axios';
import { Form, Memo } from '../types';
import * as memoApi from '../lib/api/memo';
import {
  createAsyncSaga,
  createRequestActionNames,
} from '../utils/createRequestSaga';

const [SAVE_MEMO, SAVE_MEMO_SUCCESS, SAVE_MEMO_FAILURE] =
  createRequestActionNames('memo/SAVE_MEMO');
const SET_FORM = 'memo/SET_FORM';

export const saveMemoAsync = createAsyncAction(
  SAVE_MEMO,
  SAVE_MEMO_SUCCESS,
  SAVE_MEMO_FAILURE
)<MemoState['form'], [Memo, AxiosResponse], Error>();

export const setForm = createAction(
  SET_FORM,
  (payload: MemoState['form']) => payload
)();

export const saveMemoSaga = createAsyncSaga(saveMemoAsync, memoApi.saveMemo);

export function* memoSaga() {
  yield takeLatest(SAVE_MEMO, saveMemoSaga);
}

const actions = {
  ...saveMemoAsync,
  setForm,
};

type MemoAction = ActionType<typeof actions>;

export interface MemoState {
  form: Form;
  memo: Memo;
  memoError: Error | null;
}

export const initialState: MemoState = {
  form: {
    title: '',
    content: '',
  },
  memo: {
    id: '',
    title: '',
  },
  memoError: null,
};

const memo = createReducer<MemoState, MemoAction>(initialState)
  .handleAction(setForm, (state, { payload: form }) => ({
    ...state,
    form,
  }))
  .handleAction(saveMemoAsync.success, (state, { payload: memoData }) => ({
    ...state,
    memo: memoData,
  }))
  .handleAction(saveMemoAsync.failure, (state, { payload: memoError }) => ({
    ...state,
    memoError,
  }));

export default memo;
