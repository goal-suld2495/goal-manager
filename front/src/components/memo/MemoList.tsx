import { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useRecoilValue, useSetRecoilState } from 'recoil';
import styled from 'styled-components';
import { getMemoListState, memoListStateTrigger } from '../../modules/memo';
import { Memo } from '../../types';

type MemoItemProps = {
  memo: Memo
};

const MemoListBlock = styled.ul``;

const MemoItem = ({ memo }: MemoItemProps) => {
  return (
    <Link to={`/memos/${memo.id}`}>{memo.title}</Link>
  );
};

const MemoList = () => {
  const memoList = useRecoilValue(getMemoListState);
  const setMemoListStateTrigger = useSetRecoilState(memoListStateTrigger);

  useEffect(() => {
    setMemoListStateTrigger((trigger) => {
      return trigger + 1;
    });
  }, []);

  return (
    <MemoListBlock>
      {memoList.length ? (
        memoList.map((memo) => (
          <MemoItem memo={memo} key={memo.id} />
        ))
      ) : (
        <Link to="/write">메모 생성</Link>
      )}
    </MemoListBlock>
  );
};

export default MemoList;
