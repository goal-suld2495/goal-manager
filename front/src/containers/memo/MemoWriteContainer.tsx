import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router';
import { message } from 'antd';
import MemoForm from '../../components/memo/MemoForm';
import { saveMemoAsync, setForm } from '../../modules/memo';
import { RootState } from '../../modules/store';

const MemoWriteContainer = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const {
    form,
    memo: result,
    memoError,
  } = useSelector(({ memo }: RootState) => ({
    form: memo.form,
    memo: memo.memo,
    memoError: memo.memoError,
  }));

  const onSubmit = () => {
    if (!form.title || !form.content) {
      message.warning('제목 또는 내용을 입력하세요.');
      return;
    }

    dispatch(saveMemoAsync.request(form));
  };

  const onChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    dispatch(
      setForm({
        ...form,
        [event.target.name]: event.target.value,
      })
    );
  };

  useEffect(() => {
    if (memoError) {
      message.error('저장에 실패하였습니다.');
    }
  }, [memoError]);

  useEffect(() => {
    navigate(`/memos/${result.id}`);
  }, [result]);

  return <MemoForm onSubmit={onSubmit} onChange={onChange} form={form} />;
};

export default MemoWriteContainer;
