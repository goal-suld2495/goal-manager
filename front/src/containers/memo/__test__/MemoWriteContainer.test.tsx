import { screen, waitFor } from '@testing-library/react';
import { mocked } from 'jest-mock';
import userEvent from '@testing-library/user-event';
import MemoWriteContainer from '../MemoWriteContainer';
import renderWithRedux from '../../../utils/test/renderWithRedux';
import * as memoAPI from '../../../lib/api/memo';

jest.mock('../../../lib/api/memo');
const mockedMemoApi = mocked(memoAPI, true);

describe('MemoWriteContainer', () => {
  const setup = () => {
    return renderWithRedux(<MemoWriteContainer />);
  };

  it('render UI', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('생성 버튼을 클릭했을 때, 제목 또는 내용을 입력하지 않은 경우 경고창을 출력한다.', () => {
    global.alert = jest.fn();
    setup();

    const button = screen.getByRole('button', { name: '생성' });
    userEvent.click(button);

    expect(global.alert).toHaveBeenCalledTimes(1);
  });

  it('성공적으로 저장하는 경우 작성한 게시글 뷰어 페이지로 이동한다', async () => {
    mockedMemoApi.saveMemo.mockImplementation(() => {
      return Promise.resolve({
        data: {
          title: '제목',
          content: '내용',
          id: '100',
        },
        status: 200,
        statusText: '',
        headers: '',
        config: {},
      });
    });

    setup();

    const titleInput = screen.getByLabelText('제목');
    const contentInput = screen.getByPlaceholderText('내용을 입력해 주세요.');

    userEvent.type(titleInput, '제목');
    userEvent.type(contentInput, '내용');

    const button = screen.getByRole('button', { name: '생성' });
    userEvent.click(button);

    await waitFor(() => {
      expect(global.location.pathname).toBe('/memos/100');
    });
  });
  // TODO
  // 저장 액션을 성공적으로 수행했다면 게시글 뷰어페이지로 이동

  // TODO
  // 저장 액션에 실패했다면 관련 된 모달 출력
});
