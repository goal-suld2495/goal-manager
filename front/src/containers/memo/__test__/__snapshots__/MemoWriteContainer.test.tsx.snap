// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoWriteContainer render UI 1`] = `
<div>
  <form
    class="sc-bdvvtL comMSI"
  >
    <div
      class="title"
    >
      <label
        for="title"
      >
        제목
      </label>
      <input
        id="title"
        name="title"
        type="text"
        value=""
      />
    </div>
    <div
      class="content"
    >
      <textarea
        name="content"
        placeholder="내용을 입력해 주세요."
      />
    </div>
    <div
      class="buttons"
    >
      <button
        type="button"
      >
        생성
      </button>
    </div>
  </form>
</div>
`;
